File: Data Structures & Algorithms/implement-stack-using-queues/submission-0.py
class MyStack:

    def __init__(self):
        self.q = deque()

    def push(self, x: int) -> None:
        self.q.appendleft(x)

    def pop(self) -> int:
        curr = self.q.popleft()
        return curr

    def top(self) -> int:
        curr = self.q[0]
        return curr

    def empty(self) -> bool:
        if self.q:
            return False
        return True

        


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: Data Structures & Algorithms/design-browser-history/submission-0.cs
public class PageNode
{
    public string Value;
    public PageNode? Next;
    public PageNode? Prev;

    public PageNode(string val)
    {
        Value = val;
    }
}

public class BrowserHistory 
{
    private PageNode? curr;

    public BrowserHistory(string homepage)
    {
        curr = new PageNode(homepage);
    }
    
    public void Visit(string url)
    {
        PageNode newPageNode = new PageNode(url)
        {
            Prev = curr
        };
        curr!.Next = newPageNode;
        curr = newPageNode;
    }
    
    public string Back(int steps) 
    {
        int i = 0;
        while (curr!.Prev != null && i != steps)
        {
            curr = curr.Prev;
            i++;
        }
        return curr.Value;
    }
    
    public string Forward(int steps) 
    {
        int i = 0;
        while (curr!.Next != null && i != steps)
        {
            curr = curr.Next;
            i++;
        }
        return curr.Value;
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.Visit(url);
 * string param_2 = obj.Back(steps);
 * string param_3 = obj.Forward(steps);
 */